ibis/expr/datatypes.py

```diff
@@ -1220,19 +1179,281 @@ def type(self):
+
+
+@infer.register(OrderedDict)
+def infer_struct(value):
+    if not value:
+        raise TypeError('Empty struct type not supported')
+    return Struct(
+        list(value.keys()),
+        list(map(infer, value.values()))
+    )
+
+
+@infer.register(dict)
+def infer_map(value):
+    if not value:
+        return Map(null, null)
+    return Map(
+        highest_precedence_dtype(list(map(infer, value.keys()))),
+        highest_precedence_dtype(list(map(infer, value.values()))),
+    )
+
+
+@infer.register(list)
+def infer_array(value):
+    if not value:
+        return Array(null)
+    return Array(highest_precedence_dtype(list(map(infer, value))))
+
+
+# TODO: infer ndarray, infer series
+
+
+@infer.register(datetime.time)
+def infer_time(value):
+    return time
+
+
+@infer.register(datetime.date)
+def infer_date(value):
+    return date
+
+
+@infer.register(datetime.datetime)
+def infer_timestamp(value):
+    return timestamp
+
+
+# TODO: infer pd.Timedelta, Datetime
+
+@infer.register(datetime.timedelta)
+def infer_interval(value):
+    return interval
+
+
+@infer.register(six.string_types)
+def infer_string(value):
+    return string
+
+
+@infer.register(builtins.float)
+def infer_floating(value):
+    return double
+
+
+@infer.register(six.integer_types + (np.integer,))
+def infer_integer(value):
+    return int_class(value)
+
+
+@infer.register(bool)
+def infer_boolean(value):
+    return boolean
+
+
+@infer.register((type(None), Null))
+def infer_null(value):
+    return null
+
+
+#  isinstance(a, np.generic) -> if true then map based on a.dtype
+
+# TODO:
+# multipledispatch infer_dtype - ala pandas api
+# multipledispatch infer_schema - list, dict, ordereddict, dataframe
+# to_pandas
+
+
+castable = Dispatcher('castable')
+
+
+@castable.register(DataType, DataType)
+def can_cast_subtype(source, target):
+    return isinstance(target, type(source))
+
+
+@castable.register(Any, DataType)
+def can_cast_any(source, target):
+    return True
+
+
+@castable.register(Null, DataType)
+def can_cast_null(source, target):
+    return target.nullable
+
+
+@castable.register(Integer, Integer)
+def can_cast_to_generic_integer(source, target):
+    return True
+
+
+@castable.register(UnsignedInteger, UnsignedInteger)
+def can_cast_unsigned_integer(source, target):
+    return target._nbytes >= source._nbytes
+
+
+@castable.register(SignedInteger, SignedInteger)
+def can_cast_signed_integers(source, target):
+    return target._nbytes >= source._nbytes
+
+
+@castable.register(Floating, Floating)
+def can_cast_floats(source, target):
+    return target._nbytes >= source._nbytes
+
+
+@castable.register(Integer, (Floating, Decimal))
+def can_upcast_integers(source, target):
+    return True
+
+
+@castable.register(Floating, Decimal)
+def can_upcast_floats(source, target):
+    return True
+
+
+@castable.register(Interval, Interval)
+def can_cast_intervals(source, target):
+    return castable(source.value_type, target.value_type)
```

Comment:
Hm, this is tricky because there are a few cases where casting between intervals becomes ambiguous, like casting anything longer than a week (that isn't a week itself) to weeks, e.g., casting 1 month to week. Some months may have more weeks than others.

Casting seconds to months is also fraught with annoyances.

I think we will need to check the units to make sure these are valid casts.